File: packages/core/src/main/kotlin/com/antoniofuture/ideeventsounds/core/eventmatcher/EventMatcher.kt
package com.antoniofuture.ideeventsounds.core.eventmatcher

import com.antoniofuture.ideeventsounds.core.config.ConfigManager
import com.antoniofuture.ideeventsounds.core.config.SoundMapping
import com.antoniofuture.ideeventsounds.core.soundplayer.SoundPlayer
import java.util.concurrent.ConcurrentHashMap

class EventMatcher {
    private val configManager = ConfigManager.instance
    private val soundPlayer = SoundPlayer()
    private val regexCache = ConcurrentHashMap<String, Regex>()

    fun matchAndPlay(eventKey: String, message: String? = null) {
        val config = configManager.loadConfig()
        if (!config.enable) {
            return
        }

        if (eventKey == "notification" && message != null) {
            matchAndPlayNotification(message)
            return
        }

        val soundMapping = configManager.getSoundMapping(eventKey) ?: return

        // 检查单个事件是否启用
        if (!soundMapping.isEnabled) {
            return
        }

        // 正则匹配逻辑
        if (soundMapping.regex.isNotEmpty() && message != null) {
            if (!matchesRegex(soundMapping.regex, message)) {
                return
            }
        }

        soundPlayer.playSound(eventKey)
    }

    private fun matchAndPlayNotification(message: String) {
        val config = configManager.loadConfig()
        val notificationMappings = config.sounds.filter { it.eventKey == "notification" && it.isEnabled }

        for (soundMapping in notificationMappings) {
            if (soundMapping.regex.isNotEmpty() && matchesRegex(soundMapping.regex, message)) {
                playSoundFromMapping(soundMapping)
                return
            }
        }
    }

    private fun playSoundFromMapping(soundMapping: SoundMapping) {
        val soundPath = soundMapping.soundPath
        try {
            if (soundPath.startsWith("preset/") || soundPath.startsWith("sounds/")) {
                val prefix = if (soundPath.startsWith("preset/")) "preset/" else "sounds/"
                val fileName = soundPath.substring(prefix.length)
                val resourcePath = "/preset/$fileName"
                val inputStream = javaClass.getResourceAsStream(resourcePath)
                if (inputStream != null) {
                    soundPlayer.playFromStream(inputStream)
                }
            } else {
                val soundFile = java.io.File(soundPath)
                if (soundFile.exists()) {
                    soundPlayer.playFromFile(soundFile)
                }
            }
        } catch (e: Exception) {
            println("[EventMatcher] Error playing sound: ${e.message}")
        }
    }

    private fun matchesRegex(pattern: String, message: String): Boolean {
        return try {
            val regex = regexCache.computeIfAbsent(pattern) { Regex(it) }
            regex.containsMatchIn(message)
        } catch (e: Exception) {
            // 正则表达式无效，跳过匹配（视为匹配成功）
            println("[EventMatcher] Invalid regex pattern: $pattern, error: ${e.message}")
            true
        }
    }

    fun testRegex(pattern: String, testMessage: String): Boolean {
        return try {
            val regex = Regex(pattern)
            regex.containsMatchIn(testMessage)
        } catch (e: Exception) {
            println("[EventMatcher] Regex test failed: ${e.message}")
            false
        }
    }
}

File: packages/vscode/src/eventMatcher/eventMatcher.ts
import { ConfigManager } from '../config/configManager';
import { SoundMapping } from '../config/types';
import { SoundPlayer } from '../soundPlayer/soundPlayer';

export class EventMatcher {
  private configManager: ConfigManager;
  private soundPlayer: SoundPlayer;
  private regexCache: Map<string, RegExp> = new Map();

  constructor() {
    this.configManager = ConfigManager.getInstance();
    this.soundPlayer = new SoundPlayer();
  }

  matchAndPlay(eventKey: string, message?: string): void {
    const config = this.configManager.loadConfig();
    if (!config.enable) {
      return;
    }

    if (eventKey === 'notification' && message) {
      this.matchAndPlayNotification(message);
      return;
    }

    const soundMapping = this.configManager.getSoundMapping(eventKey);
    if (!soundMapping) {
      return;
    }

    if (!soundMapping.isEnabled) {
      return;
    }

    if (soundMapping.regex && message) {
      if (!this.matchesRegex(soundMapping.regex, message)) {
        return;
      }
    }

    this.soundPlayer.playSound(eventKey);
  }

  private matchAndPlayNotification(message: string): void {
    const config = this.configManager.loadConfig();
    const notificationMappings = config.sounds.filter(
      s => s.eventKey === 'notification' && s.isEnabled
    );

    for (const soundMapping of notificationMappings) {
      if (soundMapping.regex && this.matchesRegex(soundMapping.regex, message)) {
        this.playSoundFromMapping(soundMapping);
        return;
      }
    }
  }

  private playSoundFromMapping(soundMapping: SoundMapping): void {
    try {
      if (soundMapping.soundPath.startsWith('preset/') || soundMapping.soundPath.startsWith('sounds/')) {
        this.soundPlayer.playSound(soundMapping.eventKey);
      } else {
        this.soundPlayer.playSound(soundMapping.eventKey);
      }
    } catch (e) {
      console.error(`[EventMatcher] Error playing sound: ${e}`);
    }
  }

  private matchesRegex(pattern: string, message: string): boolean {
    try {
      let regex = this.regexCache.get(pattern);
      if (!regex) {
        regex = new RegExp(pattern, 'i');
        this.regexCache.set(pattern, regex);
      }
      return regex.test(message);
    } catch (e) {
      console.error(`[EventMatcher] Invalid regex pattern: ${pattern}, error: ${e}`);
      return true;
    }
  }

  testRegex(pattern: string, testMessage: string): boolean {
    try {
      const regex = new RegExp(pattern, 'i');
      return regex.test(testMessage);
    } catch (e) {
      console.error(`[EventMatcher] Regex test failed: ${e}`);
      return false;
    }
  }
}
